src/query/interpreter.cpp

```diff
@@ -2319,53 +2446,74 @@ void Interpreter::Commit() {
+          using ErrorType = std::remove_cvref_t<T>;
+          if constexpr (std::is_same_v<ErrorType, storage::ReplicationError>) {
+            commit_confirmed_by_all_sync_repplicas = false;
+          } else if constexpr (std::is_same_v<ErrorType, storage::ConstraintViolation>) {
+            const auto &constraint_violation = arg;
+            auto &label_name = execution_db_accessor->LabelToName(constraint_violation.label);
+            switch (constraint_violation.type) {
+              case storage::ConstraintViolation::Type::EXISTENCE: {
+                MG_ASSERT(constraint_violation.properties.size() == 1U);
+                auto &property_name = execution_db_accessor->PropertyToName(*constraint_violation.properties.begin());
+                throw QueryException("Unable to commit due to existence constraint violation on :{}({})", label_name,
+                                     property_name);
+              }
+              case storage::ConstraintViolation::Type::UNIQUE: {
+                std::stringstream property_names_stream;
+                utils::PrintIterable(property_names_stream, constraint_violation.properties, ", ",
+                                     [&execution_db_accessor](auto &stream, const auto &prop) {
+                                       stream << execution_db_accessor->PropertyToName(prop);
+                                     });
+                throw QueryException("Unable to commit due to unique constraint violation on :{}({})", label_name,
+                                     property_names_stream.str());
+              }
+            }
+          } else {
+            static_assert(always_false_v<T>, "Missing type from variant visitor");
+          }
+        },
+        error);
   }
 
   // The ordered execution of after commit triggers is heavily depending on the exclusiveness of db_accessor_->Commit():
   // only one of the transactions can be commiting at the same time, so when the commit is finished, that transaction
   // probably will schedule its after commit triggers, because the other transactions that want to commit are still
   // waiting for commiting or one of them just started commiting its changes.
   // This means the ordered execution of after commit triggers are not guaranteed.
+  auto all_triggers_confirmed_by_all_sync_replicas = true;
   if (trigger_context && interpreter_context_->trigger_store.AfterCommitTriggers().size() > 0) {
     interpreter_context_->after_commit_trigger_pool.AddTask(
         [trigger_context = std::move(*trigger_context), interpreter_context = this->interpreter_context_,
-         user_transaction = std::shared_ptr(std::move(db_accessor_))]() mutable {
-          RunTriggersIndividually(interpreter_context->trigger_store.AfterCommitTriggers(), interpreter_context,
-                                  std::move(trigger_context));
+         user_transaction = std::shared_ptr(std::move(db_accessor_)),
```

Comment:
This is a very good remark, you are correct about the dangling reference and the workflow.

I see two solutions depending on what we functionally want:
1. The workflow is the source of truth: then we need to change the code. For instance we can run the triggers in the same thread (and not via the pool_thread)
2. The current code is correct and we want to keep it as it is. Then I don't see how we can have a guarantee to throw an exception as we'd like. In that case I would suggest we simply accept that we can't report a SYNC REPLICA failure when the user is modifying data. (Diagram would need be be adapted).

For now, I have updated the code to follow 1. to see if any other would be caught by the CI. We need to discuss and see which approach we take.

Reply:
Conclusion: we will not block and wait for all triggers to have run before returning to user.
Reason is:

- if the replication status is more important to the user, he can use the `BeforeCommit` triggers. This means it will wait for all triggers to have run before confirming the commit.

- if the user doesn't want to wait for all triggers, he can use the `AfterCommit` triggers.